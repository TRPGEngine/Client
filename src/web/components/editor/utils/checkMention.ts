import { Editor, Range } from 'slate';
import _get from 'lodash/get';
import _last from 'lodash/last';

export function checkMention(
  editor: Editor
): {
  target: Range;
  searchText: string;
} | null {
  const { selection, operations } = editor;

  if (selection && Range.isCollapsed(selection)) {
    // 处理输入时前面有@的情况
    const [start] = Range.edges(selection);
    const wordBefore = Editor.before(editor, start, { unit: 'word' });
    const before = wordBefore && Editor.before(editor, wordBefore);
    const beforeRange = before && Editor.range(editor, before, start);
    const beforeText = beforeRange && Editor.string(editor, beforeRange);
    const beforeMatch = beforeText && beforeText.match(/^@(\w+)$/);
    const after = Editor.after(editor, start);
    const afterRange = Editor.range(editor, start, after);
    const afterText = Editor.string(editor, afterRange);
    const afterMatch = afterText.match(/^(\s|$)/);

    if (beforeMatch && afterMatch) {
      return {
        target: beforeRange!,
        searchText: beforeMatch[1],
      };
    }
  }

  if (
    _get(operations, [0, 'type']) === 'insert_text' &&
    _get(operations, [0, 'text']) === '@'
  ) {
    // 处理输入@的情况
    const range = Editor.range(editor, _get(operations, [0, 'path']));
    return {
      target: range,
      searchText: '',
    };
  }

  return null;
}
