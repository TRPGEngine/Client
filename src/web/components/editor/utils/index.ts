import { Editor, Transforms, Range, Element } from 'slate';
import type { FormatType, TRPGEditor, TRPGEditorValue } from '../types';

const LIST_TYPES = ['numbered-list', 'bulleted-list'];

/**
 * 创建一个空的slate数据对象
 */
export function buildBlankInputData(): TRPGEditorValue[] {
  return [
    {
      children: [{ text: '' }],
    } as TRPGEditorValue,
  ];
}

/**
 * 获取文档开始位置
 */
export function getHeadSelection(): Range {
  return {
    anchor: { path: [0, 0], offset: 0 },
    focus: { path: [0, 0], offset: 0 },
  };
}

/**
 * 判定块级元素是否生效
 */
export const isBlockActive = (editor: TRPGEditor, format: FormatType) => {
  const [match] = Array.from(
    Editor.nodes(editor, {
      match: (n) =>
        !Editor.isEditor(n) && Element.isElement(n) && n.type === format,
    })
  );

  return !!match;
};

/**
 * 切换块级元素属性
 */
export const toggleBlock = (editor: TRPGEditor, format: FormatType) => {
  const isActive = isBlockActive(editor, format);
  const isList = LIST_TYPES.includes(format);

  Transforms.unwrapNodes(editor, {
    match: (n) =>
      !Editor.isEditor(n) &&
      Element.isElement(n) &&
      LIST_TYPES.includes(n.type),
    split: true,
  });

  Transforms.setNodes(editor, {
    type: isActive ? 'paragraph' : isList ? 'list-item' : format,
  });

  if (!isActive && isList) {
    const block = { type: format, children: [] };
    Transforms.wrapNodes(editor, block);
  }
};

/**
 * 判断标记是否生效
 */
export const isMarkActive = (editor: TRPGEditor, format: string) => {
  const marks = Editor.marks(editor);
  return marks ? marks[format] === true : false;
};

/**
 * 切换标记
 */
export const toggleMark = (editor: TRPGEditor, format: string) => {
  const isActive = isMarkActive(editor, format);

  if (isActive) {
    Editor.removeMark(editor, format);
  } else {
    Editor.addMark(editor, format, true);
  }
};
