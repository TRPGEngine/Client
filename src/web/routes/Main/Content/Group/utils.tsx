import type { GroupPanelType } from '@shared/types/panel';
import { Iconfont } from '@web/components/Iconfont';
import React from 'react';
import styled from 'styled-components';

const DefaultPanelIconContainer = styled.div`
  padding-left: 2px;
  padding-right: 2px;
`;

/**
 * 获取团面板Icon
 */
export function getGroupPanelIcon(panelType: GroupPanelType) {
  if (panelType === 'voicechannel') {
    return <Iconfont>&#xe664;</Iconfont>;
  } else if (panelType === 'note') {
    return <Iconfont>&#xe621;</Iconfont>;
  } else if (panelType === 'website') {
    return <Iconfont>&#xe651;</Iconfont>;
  } else if (panelType === 'calendar') {
    return <Iconfont>&#xe867;</Iconfont>;
  }

  return <DefaultPanelIconContainer>#</DefaultPanelIconContainer>;
}
