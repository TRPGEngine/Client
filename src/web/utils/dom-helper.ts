interface OpenPostWindowOptions {
  url: string;
  data: object;
  target?: string;
}

/**
 * 通过post方式提交一个表单(打开新窗口)
 */
export function openPostWindow({
  url,
  data,
  target = '_blank',
}: OpenPostWindowOptions) {
  const form = document.createElement('form');
  form.method = 'POST';
  form.action = url;
  form.target = target;

  for (const key in data) {
    if (Object.prototype.hasOwnProperty.call(data, key)) {
      const value = data[key];

      const hideInput = document.createElement('input');
      hideInput.type = 'text';
      hideInput.name = key;
      hideInput.value = value;
      form.appendChild(hideInput);
    }
  }

  document.body.appendChild(form);

  form.submit();
  document.body.removeChild(form);
}
