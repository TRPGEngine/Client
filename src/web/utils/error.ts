import _isString from 'lodash/isString';
import { notification } from 'antd';

/**
 * 通用的错误处理机制
 */
export function handleError(err: any, prefix: string = '') {
  if (_isString(err?.msg)) {
    notification.error({ message: prefix + err?.msg });
  } else {
    notification.error({ message: prefix + String(err) });
  }
}
