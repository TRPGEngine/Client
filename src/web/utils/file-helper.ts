/**
 * 传入一个图片文件， 返回对应的 Base64 编码
 * @param img 图片文件
 */
export function fileToDataUrl(img: File): Promise<string> {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.addEventListener('load', () => resolve(reader.result?.toString()));
    reader.addEventListener('error', () => reject(reader.error));
    reader.readAsDataURL(img);
  });
}

/**
 * 根据一个BlobUrl地址获取相应的Blob对象
 * @param blobUrl blobUrl 地址，由URL.createObjectURL方法生成
 */
export function blobFromUrl(blobUrl: string): Promise<Blob> {
  return new Promise((resolve, reject) => {
    const req = new XMLHttpRequest();
    req.responseType = 'blob';
    req.onload = () => {
      resolve(req.response);
    };
    req.onerror = () => {
      reject(req.responseText);
    };
    req.open('get', blobUrl);
    req.send();
  });
}

/**
 * Blob对象转文件对象
 * @param blob Blob对象
 * @param fileName 文件名
 */
export function blobToFile(blob: Blob, fileName: string): File {
  return new File([blob], fileName, {
    lastModified: new Date().valueOf(),
    type: blob.type,
  });
}

/**
 * blobUrl 转 文件对象
 * @param blobUrl blobUrl
 * @param fileName 文件名
 */
export async function blobUrlToFile(
  blobUrl: string,
  fileName: string = 'image.jpg'
): Promise<File> {
  const blob = await blobFromUrl(blobUrl);
  return blobToFile(blob, fileName);
}

/**
 * 下载Bloburl
 */
export async function downloadBlobUrl(blobUrl: string, fileName: string) {
  const a = document.createElement('a');
  a.href = blobUrl;
  a.download = fileName; // 这里填保存成的文件名
  a.click();
}

/**
 * 下载Blob文件
 */
export async function downloadBlob(blob: Blob, fileName: string) {
  const url = window.URL.createObjectURL(blob);
  downloadBlobUrl(url, fileName);
  URL.revokeObjectURL(url);
}
