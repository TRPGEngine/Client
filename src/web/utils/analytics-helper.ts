import type { UserInfo } from '@redux/types/user';
import { getUserName } from '@shared/utils/data-helper';
import posthog from 'posthog-js';
import _isString from 'lodash/isString';
import _get from 'lodash/get';
import Config from 'config';
import config from '@shared/project.config';

// Posthog
const token = _get(Config, 'posthog.token');
const instance = _get(Config, 'posthog.instance');

/**
 * 初始化统计分析系统
 */
export function initAnalytics() {
  if (_isString(token) && _isString(instance)) {
    posthog.init(token, {
      api_host: instance,
      autocapture: false, // 关闭autocapture以节约事件用量
      disable_session_recording: true, // 关闭自动录屏(不需要且一直报错)
    });
    posthog.register({
      environment: config.environment,
      version: config.version,
      platform: config.platform,
    });
  }
}

/**
 * 设置分析的用户信息
 * @param info 用户信息
 */
export function setAnalyticsUser(info: UserInfo) {
  posthog.identify(info.uuid, {
    name: getUserName(info),
    username: info.username,
  });
}

/**
 * 发送事件
 * @param name 事件名
 * @param properties 相关属性
 */
export function trackEvent(name: string, properties?: posthog.Properties) {
  posthog.capture(name, properties);
}

/**
 * 发送单页应用Url变更事件
 */
export function trackUrlChangeEvent() {
  posthog.capture('$pageview');
}
