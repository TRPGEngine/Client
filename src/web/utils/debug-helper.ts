import memoizeOne from 'memoize-one';
import { IS_NEW_APP, IS_DEVELOPER } from '@shared/utils/consts';
import { showToasts } from '@shared/manager/ui';

/**
 * 获取是否为测试用户
 * 方法是手动在 localStorage上设置 __isDeveloper: true
 */
export const checkIsDeveloper = memoizeOne(() => {
  return localStorage.getItem(IS_DEVELOPER) === 'true';
});

/**
 * 切换APP
 * @param toNewApp 是否切换到新版
 */
export function switchToAppVersion(toNewApp = true) {
  localStorage.setItem(IS_NEW_APP, String(toNewApp));

  showToasts('切换成功, 1秒后回到首页');
  setTimeout(() => {
    window.location.href = '/'; // 刷新页面并回到首页
  }, 1000);
}
