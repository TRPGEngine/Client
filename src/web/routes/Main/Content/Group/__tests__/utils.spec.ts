import type { GroupPanelType } from '@shared/types/panel';
import { getGroupPanelIcon } from '../utils';

describe('getGroupPanelIcon', () => {
  test.each<[GroupPanelType]>([
    ['channel'],
    ['voicechannel'],
    ['note'],
    ['website'],
    ['calendar'],
  ])('%s', (type) => {
    expect(getGroupPanelIcon(type)).toMatchSnapshot();
  });
});
