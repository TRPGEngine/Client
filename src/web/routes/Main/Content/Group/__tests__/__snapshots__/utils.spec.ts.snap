// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGroupPanelIcon calendar 1`] = `
<Iconfont>
  
</Iconfont>
`;

exports[`getGroupPanelIcon channel 1`] = `
<styled.div>
  #
</styled.div>
`;

exports[`getGroupPanelIcon note 1`] = `
<Iconfont>
  
</Iconfont>
`;

exports[`getGroupPanelIcon voicechannel 1`] = `
<Iconfont>
  
</Iconfont>
`;

exports[`getGroupPanelIcon website 1`] = `
<Iconfont>
  
</Iconfont>
`;
